packages/protocol/contracts/liquidity/GrandaMento.sol

```diff
@@ -0,0 +1,477 @@
+pragma solidity ^0.5.13;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+
+import "../common/FixidityLib.sol";
+import "../common/InitializableV2.sol";
+import "../common/linkedlists/IntegerLinkedList.sol";
+import "../common/linkedlists/LinkedList.sol";
+import "../common/UsingRegistry.sol";
+import "../common/interfaces/ICeloVersionedContract.sol";
+import "../common/libraries/ReentrancyGuard.sol";
+import "../stability/interfaces/IStableToken.sol";
+
+/**
+ * @title Facilitates large exchanges between CELO stable tokens.
+ */
+contract GrandaMento is
+  ICeloVersionedContract,
+  Ownable,
+  InitializableV2,
+  UsingRegistry,
+  ReentrancyGuard
+{
+  using FixidityLib for FixidityLib.Fraction;
+  using IntegerLinkedList for LinkedList.List;
+  using SafeMath for uint256;
+
+  // Emitted when a new exchange proposal is created.
+  event ExchangeProposalCreated(
+    uint256 indexed proposalId,
+    address indexed exchanger,
+    address indexed stableToken,
+    uint256 sellAmount,
+    uint256 buyAmount,
+    bool sellCelo
+  );
+
+  // Emitted when an exchange proposal is approved by the approver.
+  event ExchangeProposalApproved(uint256 indexed proposalId);
+
+  // Emitted when an exchange proposal is cancelled.
+  event ExchangeProposalCancelled(uint256 indexed proposalId, address sender);
+
+  // Emitted when an exchange proposal is executed.
+  event ExchangeProposalExecuted(uint256 indexed proposalId);
+
+  // Emitted when the approver is set.
+  event ApproverSet(address approver);
+
+  // Emitted when the spread is set.
+  event SpreadSet(uint256 spread);
+
+  // Emitted when the veto period in seconds is set.
+  event VetoPeriodSecondsSet(uint256 vetoPeriodSeconds);
+
+  // Emitted when the exchange limits for a stable token are set.
+  event StableTokenExchangeLimitsSet(
+    string stableTokenRegistryId,
+    uint256 minExchangeAmount,
+    uint256 maxExchangeAmount
+  );
+
+  enum ExchangeProposalState { None, Proposed, Approved, Executed, Cancelled }
+
+  struct ExchangeLimits {
+    // The minimum amount of an asset that can be exchanged in a single proposal.
+    uint256 minExchangeAmount;
+    // The maximum amount of an asset that can be exchanged in a single proposal.
+    uint256 maxExchangeAmount;
+  }
+
+  struct ExchangeProposal {
+    // The exchanger/proposer of the exchange proposal.
+    address payable exchanger;
+    // The stable token involved in this proposal.
+    address stableToken;
```

Comment:
This one's actually important to use the address rather than the ID just in case a stable token -> CELO proposal is created, the exchanger deposits that stable token, and then the address in the registry of the stable token changes. Then when the exchange is cancelled or executed, the contract will try to burn / transfer the deposited tokens from the new address rather than the old one

Reply:
hmm
I think we would never want to make a storage migration where the new balances don't reflect the old balances

Reply:
Given that, how do you feel about the assumption the token address will never change?

Reply:
I have imagined a pattern where we could have a new address but the new contract falls back to the old contract's state if uninitialized/migrated
I don't feel strongly about this case though

Reply:
Chatted 1:1, keeping with this approach